design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
+
+We may want to introduce a limit to the number of rows that may be returned from these `executeQuery()` methods,
+in order to decrease the likelihood of running out of memory.
+
+### Insert values
+
+A method to insert one row at a time:
+```
+  /**
+   * Insert a single row into the relevant stream/table.
+   *
+   * @param streamName name of stream/table.
+   * @param row the row to insert.
+   * @return a future that completes once the request has been processed.
+   */
+  CompletableFuture<Void> insertInto(String streamName, Map<String, Object> row);
+```
+
+A method to stream inserts (via the `/inserts-stream` endpoint):
+```
+  CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<List<Object>> insertsPublisher);
+}
+```
+where `InsertResponse` is as follows:
+```
+public interface InsertResponse {
+
+  /**
+   * Whether the row was successfully inserted or not.
+   */
+  boolean isSuccessful();
+
+  /**
+   * Unique sequence number for the row in the stream of inserts.
+   */
+  int getSequenceNum();
+
+  /**
+   * If unsuccessful, the error message.
+   */
+  String getErrorMessage();
+
+  /**
+   * If unsuccessful, the error code.
+   */
+  int getErrorCode();
+
+}
+```
+
+### DDL/DML operations
+
+#### `CREATE <STREAM/TABLE>`, `CREATE <STREAM/TABLE> ... AS SELECT`, `INSERT INTO`
+
+`Client` methods:
+```
+  /**
+   * Execute DDL statement: `CREATE STREAM`, `CREATE TABLE`
+   */
+  CompletableFuture<DdlResponse> executeDdl(String sql);
```

Comment:
why do we need two different apis for DDL and DML? why not just `executeStatement(String sql)` - the response object can have different types if we need to disambiguate the responses

Reply:
I'm fine with combining the DDL and DML statements but my concern with `executeStatement()` is that users might be tempted to issue other kinds of statements (create connector, admin operations, etc) with the method which wouldn't work since those need different return types.

Reply:
I think it's fine to document the intended use case and then throw an exception if an unexpected statement comes in, but that's just an opinion

Reply:
You think the client should be in the business of trying to parse statements before sending them to the server? Or are you suggesting the client throw an error if the entity it receives back from the server isn't what it was expecting.

Reply:
I feel like we have the same problem either way, no? You can't stop them from submitting DDLs to the DML method. I'm unsure about whether the client should be in the business of parsing or not - I see arguments both ways.

This is the problem with accepting string SQL in the API :/ it would've been nice to have structured objects instead of strings, but I understand that this is something standard (e.g. in JDBC)